### PR TITLE
Ported the GraphViz `dot`-format output for query graphs

### PR DESCRIPTION
**Pull request by duckki**
added query_graph::output::to_dot function
- outputs query graphs in the GraphViz Dot format
- supports both federated schemas and non-federated ones

### CODE REVIEW

**Comment by duckki** [comment]
I didn't add CLI, yet. But, I can added the `querygraph` and `federatedgraph` CLI commands once https://github.com/apollographql/federation-next/pull/246 merges.

**Review by goto-bus-stop — COMMENTED** [review]
Nice, implementation looks good. I think most of the `dot_str.push_str(&format!())` calls would read better with `write!(&mut dot_str, etc)` (or `writeln!()` as appropriate). I have some rust comments for the CLI part.

File: cli/src/main.rs [hunk 1]
```diff
@@ -17,27 +22,98 @@ enum Command {
...
+    if file_paths.len() == 0 {
+        panic!("Error: missing command arguments");
+    } else if file_paths.len() == 1 {
+        let doc_str = std::fs::read_to_string(&file_paths[0]).unwrap();
+        apollo_federation::Supergraph::new(&doc_str)
+    } else {
+        let schemas: Vec<_> = file_paths
+            .iter()
+            .map(|pathname| {
+                let doc_str = std::fs::read_to_string(&pathname).unwrap();
```
[the remark this answers is not in the document]
Comment (duckki) [reply cli/src/main.rs]: I don't think it's necessary for this CLI. But, I'll look into anyhow and use it where ever appropriate!

**Comment by duckki** [comment]
@goto-bus-stop Thanks for the review! I just pushed a new commit. Please review one more time.

File: src/query_graph/output.rs [hunk 2]
```diff
@@ -0,0 +1,149 @@
...
+            &cluster_dot_config,
+            &(|_, er| label_edge(er.weight())),
+            &(|_, (_, node)| label_cluster_node(node)),
+        )
+        .to_string();
+
+        writeln!(dot_str, r#"  subgraph "cluster_{}" {{"#, &cluster_name)?;
+        writeln!(dot_str, r#"    label = "Subgraph \"{}\"";"#, &cluster_name)?;
+        writeln!(dot_str, r#"    color = "black";"#)?;
+        writeln!(dot_str, r#"    style = """#)?;
```
Comment (goto-bus-stop) [inline src/query_graph/output.rs:118]: I'm not too familiar with dot syntax, but does this need a semicolon?
```suggestion
        writeln!(dot_str, r#"    style = "";"#)?;
```

File: cli/src/main.rs [hunk 3]
```diff
@@ -17,27 +22,91 @@ enum Command {
...
+        panic!("Error: missing command arguments");
+    } else if file_paths.len() == 1 {
+        let doc_str = std::fs::read_to_string(&file_paths[0]).unwrap();
+        apollo_federation::Supergraph::new(&doc_str)
+    } else {
+        let schemas: Vec<_> = file_paths
+            .iter()
+            .map(|pathname| {
+                let doc_str = std::fs::read_to_string(pathname).unwrap();
+                let url = "file://".to_string() + pathname.to_str().unwrap();
```
Comment (goto-bus-stop) [inline cli/src/main.rs:83]: Ah, missed this the first time, it's almost always better to use the `format!()` macro than the string `+` operator, it reads a bit better and requires fewer memory allocations.
```suggestion
                let url = format!("file://{}", pathname.to_str().unwrap());
```

File: src/query_graph/output.rs [hunk 4]
```diff
@@ -0,0 +1,149 @@
...
+        cluster_name: &NodeStr,
+        edge_index: EdgeIndex,
+    ) -> bool {
+        match graph.edge_endpoints(edge_index) {
+            None => false,
+
+            Some((n1, n2)) => {
+                graph[n1].source == *cluster_name && graph[n2].source == *cluster_name
+            }
+        }
```
Comment (goto-bus-stop) [inline src/query_graph/output.rs:59]: This is perfectly fine but you can also use the [`.is_some_and`](https://doc.rust-lang.org/stable/std/option/enum.Option.html#method.is_some_and) method here:
```suggestion
        graph.edge_endpoints(edge_index).is_some_and(|(n1, n2)| {
            graph[n1].source == *cluster_name && graph[n2].source == *cluster_name
        })
```

File: src/query_graph/output.rs (hunk 2 above)
Comment (goto-bus-stop) [reply src/query_graph/output.rs:118]: Ah, it looks like it's optional.

**Review by goto-bus-stop — APPROVED** [review]
LGTM, some more stylistic nits if you feel like it, but no worries :)

File: src/query_graph/output.rs (hunk 2 above)
Comment (duckki) [reply src/query_graph/output.rs:118]: I believe so. I tried mimic the JS implementation's behavior :)

File: src/query_graph/output.rs [hunk 5]
```diff
@@ -0,0 +1,149 @@
...
+    fn edge_across_clusters(graph: &StableInnerGraph, edge_index: EdgeIndex) -> bool {
+        match graph.edge_endpoints(edge_index) {
+            None => false,
+
+            Some((n1, n2)) => graph[n1].source != graph[n2].source,
+        }
+    }
+
+    fn label_cluster_node(node: &QueryGraphNode) -> String {
+        format!(r#"label="{}@{}""#, node.type_, node.source)
```
[the remark this answers is not in the document]
Comment (duckki) [reply src/query_graph/output.rs:71]: makes sense. will do!

**Pull request merged**